geoh5py/shared/cut_by_extent.py

```diff
@@ -0,0 +1,518 @@
+#  it under the terms of the GNU Lesser General Public License as published by '
+#  the Free Software Foundation, either version 3 of the License, or           '
+#  (at your option) any later version.                                         '
+#                                                                              '
+#  geoh5py is distributed in the hope that it will be useful,                  '
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of              '
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the               '
+#  GNU Lesser General Public License for more details.                         '
+#                                                                              '
+#  You should have received a copy of the GNU Lesser General Public License    '
+#  along with geoh5py.  If not, see <https://www.gnu.org/licenses/>.           '
+# ''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
+from __future__ import annotations
+
+from dataclasses import dataclass
+
+import numpy as np
+
+from geoh5py.shared.utils import clean_extent_for_intersection
+
+
+@dataclass
+class Plane:
+    """
+    Represents a 3D plane with origin and basis vectors.
+
+    :param origin: Origin point of the plane coordinate system.
+    :param u_vector: U-axis basis vector.
+    :param v_vector: V-axis basis vector.
+    :param normal: Normal vector of the plane.
+    """
+
+    origin: np.ndarray
+    u_vector: np.ndarray
+    v_vector: np.ndarray
+    normal: np.ndarray
+
+    @classmethod
+    def from_quad_origin_last(cls, vertices: np.ndarray) -> Plane:
+        """
+        Create a coordinate system from a quad with the last vertex as origin.
+
+        Constructs a plane coordinate system (origin, u-vector, v-vector, normal)
+        from four vertices where the last vertex serves as the origin.
+
+        :param vertices: Array of shape (4, 3) containing the four vertices of the quad.
+
+        :return: Plane object containing origin and basis vectors.
+
+        :raises ValueError: If vertices is not a (4, 3) array.
+        """
+        if vertices.shape != (4, 3):
+            raise ValueError("vertices must be a (4, 3) array")
+
+        vertices = np.asarray(vertices)
+
+        origin = vertices[3]
+        u_vector = normalize(vertices[2] - origin)
+        v_vector = normalize(vertices[0] - origin)
+        normal = normalize(np.cross(u_vector, v_vector))
+        return cls(origin=origin, u_vector=u_vector, v_vector=v_vector, normal=normal)
+
+
+def normalize(vector: np.ndarray) -> np.ndarray:
+    """
+    Normalize a vector to unit length.
+
+    :param vector: Input vector to normalize.
+    :param dtype: NumPy data type for the normalized vector.
+
+    :return: Normalized vector with unit length.
+    """
+    vector = np.asarray(vector, dtype=np.float64)
+    return vector / np.linalg.norm(vector)
+
+
+def intersect_segment_plane(
```

Comment:
all the functions feel like they should be methods of the Plane class, any reason to keep it a dataclass?